### DISCUSSION TITLE
Checkout a branch from a fork

### DISCUSSION THREAD

**Discussion by Ben3eeE** [community/community#23445] [New to GitHub]
<p>A common thing that I do is to test out pull requests to ensure that the proposed changes are working and don’t introduce any regressions before merging them. My workflow for doing this is often to just checkout the specific branch and start Atom in dev mode since I already have many of the packages linked in Atom or a fresh build of master. However when someone from the community opens a pull request from their fork of one of our repositories I always clone their fork to another directory and in the case of core packages I relink this to Atom. This eventually causes me to have a mess of folders all for the same repository and there are quite many steps involved to get it up and running for testing. For pull requests to atom/atom this takes a long time because it involves rebuilding Atom entirely in the new clone.</p>
<p>Is there some way I could improve my workflow to checkout the branch directly to the clone I already have even when the branch is created on a fork so I don’t have to clone every time I want to test out a PR?</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>This is a great question!</p>
<p>For clarity, you’re wanting to test a PR to a GitHub repository that you own or maintain, let’s call it ben3eee/some-repo. That PR was submitted by someone who does not have push access to the repository, so they submitted the PR from a fork of the repository that you own or maintain. We’ll call that fork contributor/some-repo. You want to take a look at their changes in your local copy of ben3eee/some-repo as a branch.</p>
<p>First of all, you can find the most up-to-date information about this in the GitHub Help under <a href="https://help.github.com/articles/checking-out-pull-requests-locally/">Checking out pull requests locally</a>. But for a little more background, when someone submits a PR, a <a href="https://git-scm.com/book/en/v2/Git-Internals-Git-References">Git reference</a> (or “ref”) is created that stores the reference to those commits in the repository on GitHub. You can see this if you run the following in your local repo:</p>
<pre><code>git ls-remote --refs origin
</code></pre>
<p>You’ll see all of the refs on the remote named origin, including some number of refs in the format refs/pull/123/head. In the case of refs/pull/123/head, that is a reference to the head of the branch for PR <span class="hashtag">#123</span> in the repository. Knowing that GitHub stores that information this way, you can fetch that ref to your local repository and check out the branch with the following command:</p>
<pre><code>git fetch origin pull/123/head:pr/123 &amp;&amp; git checkout pr/123
</code></pre>
<p>You will then be on a local branch named pr/123 containing the code in PR <span class="hashtag">#123</span>. For my personal use, I’ve created <a href="https://github.com/lee-dohm/dotfiles/blob/8d3c59004154571578c2b32df2cdebb013517630/gitconfig#L8">an alias in my Git configuration</a> to make the process simpler so I can type:</p>
<pre><code>git co-pr 123
</code></pre>
<p>to achieve the same effect.</p>
<hr>
<p>Since I wrote the above over two years ago, GitHub has since released the GitHub CLI tool:</p>
<aside class="onebox whitelistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/cli/cli" target="_blank">GitHub</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://repository-images.githubusercontent.com/212613049/56c08980-4ce2-11ea-98b6-9d3d404437e5" class="thumbnail" width="690" height="362"></div>

<h3><a href="https://github.com/cli/cli" target="_blank">cli/cli</a></h3>

<p>GitHub’s official command line tool. Contribute to cli/cli development by creating an account on GitHub.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>When using the GitHub CLI, the same process above can be achieved using the command:</p>
<pre><code class="lang-shell">gh pr checkout 123
</code></pre>

**Comment by Ben3eeE** [comment]
<p>Thank you so much for this answer :heart: The alias in the git configuration looks great!</p>

**Comment by ShekharReddy4** [comment]
<p>Thanks, <a class="mention" href="/lee-dohm">@lee-dohm</a>. This is very helpful for project maintainers. By the way, could you also provide any best practices regarding the branching a maintainer should follow (viz contributors to submit PR’s etc.)? Please provide the practices that are widely used on GitHub. :smiley:</p>

**Comment by lee-dohm** [comment]
<p>The best practices that we recommend is summarized as <a href="https://guides.github.com/introduction/flow/" rel="nofollow noopener">the GitHub Flow</a>. We also created <a href="https://guides.github.com/introduction/flow/" rel="nofollow noopener">an illustrated guide</a> that many find helpful. Beyond that, different projects have different needs and there is a myriad of branching strategies as a result. They all are based on the common theme of the GitHub Flow though, so that’s where we recommend people start and experiment from there :+1:</p>

**Comment by GarrettVD** [comment]
<p>macOS version of the above alias:</p>
<pre><code>copr() { git fetch origin "refs/pull/$1/head:pr/$1" &amp;&amp; git checkout "pr/$1"; }
alias co-pr="copr"
</code></pre>

**Comment by meetKazuki** [comment]
<p>Can I push to that reference?</p>

**Comment by Arcanorum** [comment]
<p>How to do this but also set tracking, so we can pull any updates to the checked out branch on the fork?</p>

**Comment by nanisadw3** [comment]
You can check out a branch from a fork using either the standard Git CLI or the GitHub CLI (`gh`):
  
  ### Option 1: Using GitHub CLI (Fastest)
  If it's an open Pull Request from the fork:
  ```bash
  gh pr checkout <PR_NUMBER>
  ```
  
  ### Option 2: Using standard Git
  1. Add the fork as a new remote:
  ```bash
  git remote add fork-user https://github.com/fork-user/repository.git
  ```
  2. Fetch branches from the fork:
  ```bash
  git fetch fork-user
  ```
  3. Check out the remote branch into a local branch:
  ```bash
  git checkout -b <local-branch-name> fork-user/<branch-name>
  ```

  If this helped you check out the branch, please feel free to mark this as the accepted answer!
